{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e4b5695-be2c-4605-bc44-b0935fad2698",
   "metadata": {},
   "outputs": [],
   "source": [
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "49694e99-cd38-461f-907b-0ff891d2aa84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting selenium\n",
      "  Downloading selenium-4.35.0-py3-none-any.whl.metadata (7.4 kB)\n",
      "Collecting urllib3<3.0,>=2.5.0 (from urllib3[socks]<3.0,>=2.5.0->selenium)\n",
      "  Downloading urllib3-2.5.0-py3-none-any.whl.metadata (6.5 kB)\n",
      "Collecting trio~=0.30.0 (from selenium)\n",
      "  Downloading trio-0.30.0-py3-none-any.whl.metadata (8.5 kB)\n",
      "Collecting trio-websocket~=0.12.2 (from selenium)\n",
      "  Downloading trio_websocket-0.12.2-py3-none-any.whl.metadata (5.1 kB)\n",
      "Requirement already satisfied: certifi>=2025.6.15 in /opt/anaconda3/lib/python3.12/site-packages (from selenium) (2025.7.14)\n",
      "Collecting typing_extensions~=4.14.0 (from selenium)\n",
      "  Downloading typing_extensions-4.14.1-py3-none-any.whl.metadata (3.0 kB)\n",
      "Requirement already satisfied: websocket-client~=1.8.0 in /opt/anaconda3/lib/python3.12/site-packages (from selenium) (1.8.0)\n",
      "Collecting attrs>=23.2.0 (from trio~=0.30.0->selenium)\n",
      "  Downloading attrs-25.3.0-py3-none-any.whl.metadata (10 kB)\n",
      "Requirement already satisfied: sortedcontainers in /opt/anaconda3/lib/python3.12/site-packages (from trio~=0.30.0->selenium) (2.4.0)\n",
      "Requirement already satisfied: idna in /opt/anaconda3/lib/python3.12/site-packages (from trio~=0.30.0->selenium) (3.7)\n",
      "Collecting outcome (from trio~=0.30.0->selenium)\n",
      "  Downloading outcome-1.3.0.post0-py2.py3-none-any.whl.metadata (2.6 kB)\n",
      "Requirement already satisfied: sniffio>=1.3.0 in /opt/anaconda3/lib/python3.12/site-packages (from trio~=0.30.0->selenium) (1.3.0)\n",
      "Collecting wsproto>=0.14 (from trio-websocket~=0.12.2->selenium)\n",
      "  Downloading wsproto-1.2.0-py3-none-any.whl.metadata (5.6 kB)\n",
      "Requirement already satisfied: pysocks!=1.5.7,<2.0,>=1.5.6 in /opt/anaconda3/lib/python3.12/site-packages (from urllib3[socks]<3.0,>=2.5.0->selenium) (1.7.1)\n",
      "Requirement already satisfied: h11<1,>=0.9.0 in /opt/anaconda3/lib/python3.12/site-packages (from wsproto>=0.14->trio-websocket~=0.12.2->selenium) (0.14.0)\n",
      "Downloading selenium-4.35.0-py3-none-any.whl (9.6 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m9.6/9.6 MB\u001b[0m \u001b[31m1.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m0m\n",
      "\u001b[?25hDownloading trio-0.30.0-py3-none-any.whl (499 kB)\n",
      "Downloading trio_websocket-0.12.2-py3-none-any.whl (21 kB)\n",
      "Downloading typing_extensions-4.14.1-py3-none-any.whl (43 kB)\n",
      "Downloading urllib3-2.5.0-py3-none-any.whl (129 kB)\n",
      "Downloading attrs-25.3.0-py3-none-any.whl (63 kB)\n",
      "Downloading outcome-1.3.0.post0-py2.py3-none-any.whl (10 kB)\n",
      "Downloading wsproto-1.2.0-py3-none-any.whl (24 kB)\n",
      "Installing collected packages: wsproto, urllib3, typing_extensions, attrs, outcome, trio, trio-websocket, selenium\n",
      "  Attempting uninstall: urllib3\n",
      "    Found existing installation: urllib3 2.2.3\n",
      "    Uninstalling urllib3-2.2.3:\n",
      "      Successfully uninstalled urllib3-2.2.3\n",
      "  Attempting uninstall: typing_extensions\n",
      "    Found existing installation: typing_extensions 4.11.0\n",
      "    Uninstalling typing_extensions-4.11.0:\n",
      "      Successfully uninstalled typing_extensions-4.11.0\n",
      "  Attempting uninstall: attrs\n",
      "    Found existing installation: attrs 23.1.0\n",
      "    Uninstalling attrs-23.1.0:\n",
      "      Successfully uninstalled attrs-23.1.0\n",
      "Successfully installed attrs-25.3.0 outcome-1.3.0.post0 selenium-4.35.0 trio-0.30.0 trio-websocket-0.12.2 typing_extensions-4.14.1 urllib3-2.5.0 wsproto-1.2.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    " pip install selenium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "67866568-f729-4984-8c5f-afb4de36858d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: xlrd in /opt/anaconda3/lib/python3.12/site-packages (2.0.1)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install xlrd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "492880c4-cff6-4c57-9a61-36184d654d4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "import time\n",
    "import re\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import datetime "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca70bfac",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fec2d96c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3306f167-4b1b-4b4f-92fb-d400a58ef680",
   "metadata": {},
   "source": [
    "# Load the Excel file\n",
    "file_path = '/Users/aswathy.n/Desktop/QuickRegressionTesting1.xlsx'\n",
    "url_sheet_name = 'URLSheet'\n",
    "data_sheet_name = 'DataSheet'\n",
    "\n",
    "# Read the URL from the first sheet\n",
    "df_url = pd.read_excel(file_path, sheet_name=0)\n",
    "url_column = 'URL'\n",
    "base_url = df_url.at[0, url_column]\n",
    "\n",
    "# Read the input texts and expected responses from the specified sheet\n",
    "df_data = pd.read_excel(file_path, sheet_name=data_sheet_name)\n",
    "skuCol = 'SKU'\n",
    "questionCol = 'Question'\n",
    "expected_response_column = 'Expected Response'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "84bf9dbe-226c-4e41-a186-41f36f0889c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sheet names: URLSheet\n",
      "First sheet data:\n",
      "(3, 1)\n",
      "Second sheet data:\n",
      "                                            Question      SKU  \\\n",
      "0                                           dilution    34577   \n",
      "1  What are the different gel formulations availa...  NP00061   \n",
      "2                         is this product hazardous?   NP0006   \n",
      "3  what should I do if this accidentally gets in ...     B153   \n",
      "4                                Can i consume this?     B153   \n",
      "\n",
      "                                     Expected Answer  \n",
      "0  Recommended Dilution Information for SuperSign...  \n",
      "1  NuPAGE® Novex® Pre-Cast Gel Formulations\\nThe ...  \n",
      "2  Hazard Information for NuPAGE Transfer Buffer ...  \n",
      "3  I was unable to find specific information on w...  \n",
      "4  Consumption of Bis-ANS\\nBis-ANS is intended fo...  \n"
     ]
    }
   ],
   "source": [
    "# Load the Excel file\n",
    "excel_file = pd.ExcelFile('/Users/aswathy.n/Desktop/QuickRegressionTesting5.xlsx')\n",
    "\n",
    "# List all sheet names\n",
    "sheet_names = excel_file.sheet_names[0]\n",
    "print(\"Sheet names:\", sheet_names)\n",
    "\n",
    "# Read the first sheet into a DataFrame\n",
    "df_sheet1 = pd.read_excel(excel_file, sheet_name='URLSheet')\n",
    "print(\"First sheet data:\")\n",
    "print(df_sheet1.shape)\n",
    "pdp_AIWidget_Prod=df_sheet1.loc[0,'URL']\n",
    "\n",
    "# Read the second sheet into a DataFrame\n",
    "df_sheet2 = pd.read_excel(excel_file, sheet_name='Prod-PDPAIWidget')\n",
    "print(\"Second sheet data:\")\n",
    "print(df_sheet2.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6086c2b3-a251-417e-8b0b-c8750f59b27c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://www.thermofisher.com/order/catalog/product/NP00061?SID=srch-srp-NP00061'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdp_AIWidget_Prod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "af464b19-fb08-4baf-844c-e189475fe5fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current date time is 22:05:16.063160\n"
     ]
    }
   ],
   "source": [
    "import datetime\n",
    "now=datetime.datetime.now()\n",
    "print (\"current date time is \"+ str(now.time()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "c7e35f42-851b-414d-8bc6-09bbce976bba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize the web driver (make sure to have the appropriate driver for your browser)\n",
    "driver = webdriver.Chrome() #executable_path='/Users/aswathy.n/Documents/chromedriver')\n",
    "driver.maximize_window()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcb3d69e-3e4b-4fe4-97c0-2cd6ff4edd19",
   "metadata": {},
   "source": [
    "#This is for hard coded run\n",
    "# Initialize lists to store results\n",
    "matches = []\n",
    "mismatches = []\n",
    "\n",
    "# Function to calculate cosine similarity\n",
    "def calculate_cosine_similarity(text1, text2):\n",
    "    vectorizer = TfidfVectorizer().fit_transform([text1, text2])\n",
    "    vectors = vectorizer.toarray()\n",
    "    cosine_sim = cosine_similarity(vectors)\n",
    "    return cosine_sim[0][1]\n",
    "\n",
    "# Function to modify the URL\n",
    "def modify_url(base_url, new_value):\n",
    "    modified_url = re.sub(r'/product/[^?]+', f'/product/{new_value}', base_url)\n",
    "    \n",
    "    modified_url = re.sub(r'srp-[^&]+', f'srp-{new_value}', modified_url)\n",
    "    modified_url=\"https://www.thermofisher.com/order/catalog/product/NP00061?SID=srch-srp-NP00061\"\n",
    "    return modified_url\n",
    "\n",
    "# Initialize the previous URL modifier value\n",
    "previous_url_modifier = None\n",
    "\n",
    "# Open the web portal with the modified URL\n",
    "modified_url=\"https://www.thermofisher.com/order/catalog/product/NP00061?SID=srch-srp-NP00061\"\n",
    "driver.get(modified_url)\n",
    "\n",
    "\n",
    "time.sleep(10)\n",
    "\n",
    "cookie_accept=driver.find_element(By.ID, 'truste-consent-button')\n",
    "cookie_accept.click()\n",
    "time.sleep(15)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "048ecdbe-9adb-4d70-875d-3b347d271d86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize lists to store results\n",
    "matches = []\n",
    "mismatches = []\n",
    "\n",
    "\n",
    "# Function to calculate cosine similarity\n",
    "def calculate_cosine_similarity(text1, text2):\n",
    "    vectorizer = TfidfVectorizer().fit_transform([text1, text2])\n",
    "    vectors = vectorizer.toarray()\n",
    "    cosine_sim = cosine_similarity(vectors)\n",
    "    return cosine_sim[0][1]\n",
    "\n",
    "# Function to modify the URL\n",
    "def modify_url(base_url, new_value):\n",
    "    modified_url = re.sub(r'/product/[^?]+', f'/product/{new_value}', base_url)\n",
    "    \n",
    "    modified_url = re.sub(r'srp-[^&]+', f'srp-{new_value}', modified_url)\n",
    "   # modified_url=\"https://www.thermofisher.com/order/catalog/product/NP00061?SID=srch-srp-NP00061\"\n",
    "    print (\"modified_url\")\n",
    "    print (modified_url, new_value)\n",
    "    return modified_url\n",
    "\n",
    "# Initialize the previous URL modifier value\n",
    "previous_url_modifier = None\n",
    "\n",
    "def set_cookie(driver, name, value):\n",
    "    driver.add_cookie({'name': name, 'value': value})\n",
    "\n",
    "# Function to delete the cookie\n",
    "def delete_cookie(driver, name):\n",
    "    driver.delete_cookie(name)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "408c8e1a-c669-4ddf-9408-53e1aecd616d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open the web portal with the modified URL\n",
    "base_url=pdp_AIWidget_Prod\n",
    "driver.get(base_url)\n",
    "\n",
    "time.sleep(10)\n",
    "\n",
    "cookie_accept=driver.find_element(By.ID, 'truste-consent-button')\n",
    "cookie_accept.click()\n",
    "time.sleep(15)\n",
    "\n",
    "\n",
    "# Delete the existing cookie CK_ISO_CODE\n",
    "delete_cookie(driver, 'CK_ISO_CODE')\n",
    "#Setting new value\n",
    "set_cookie(driver, 'CK_ISO_CODE', 'US')\n",
    "\n",
    "\n",
    "driver.refresh()\n",
    "time.sleep(10)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "d11bbb79-9b7f-46a7-a13f-99074e983a3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "c930875a-f76a-4330-9d27-7f21ff71fb25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34577 Recommended Dilution Information for SuperSignal West Pico PLUS Chemiluminescent Substrate\n",
      "Primary antibody: 0.2 to 1.0 µg/mL (1:1000 to 1:5000 dilution from a 1 mg/mL stock)\n",
      "Secondary antibody: 10 to 50 ng/mL (1:20,000 to 1:100,000 dilution from 1 mg/mL stock)\n",
      "For optimal results, ensure that your western blot procedure is well-optimized, including sample amount, gel type, transfer method, membrane type, blocking reagent, wash buffer, and incubation times.\n",
      "Sources 1\n",
      "3\n",
      "5\n",
      "Additional Data & References\n",
      "1. FAQ:What is the recommended dilution range for SuperSignal West Pico PLUS Chemiluminescent Substrate?\n",
      "\n",
      "2. Antibodies for Western Blotting\n",
      "3. ECL Western Blot Substrates\n",
      "4. Chromogenic Western Blot Detection\n",
      "5. User Guide: SuperSignal West Pico PLUS Chemiluminescent Substrate (Page: 1)\n",
      "6. Chemiluminescent Western Blot Detection\n",
      "7. Brochure: Western workflow overview brochure (Chinese) (Page: 23, 45) None\n",
      "modified_url\n",
      "https://www.thermofisher.com/order/catalog/product/34577?SID=srch-srp-34577 34577\n",
      "Recommended Dilution for SuperSignal West Pico PLUS Chemiluminescent Substrate\n",
      "Primary antibody dilution: 0.2 to 1.0 \u0003bcg/mL, which corresponds to a 1:1000 to 1:5000 dilution from a 1 mg/mL stock solution.\n",
      "Secondary antibody dilution: 10 to 50 ng/mL, corresponding to a 1:20,000 to 1:100,000 dilution from a 1 mg/mL stock solution.\n",
      "These dilution ranges are recommended to optimize signal intensity and sensitivity for western blotting using this substrate.\n",
      "Additional Notes\n",
      "Optimal antibody dilutions may vary depending on the specific antibody and experimental conditions.\n",
      "It is advisable to test a range of dilutions within these recommended ranges to determine the best conditions for your assay.\n",
      "For more detailed guidance, please refer to the Tech Tip #32 Guide to Enzyme Substrates for Western Blotting and our Protein Electrophoresis and Western Blotting Support Center.\n",
      "Sources 6\n",
      "Additional Data & References\n",
      "1. ECL Western Blot Substrates\n",
      "2. Antibodies for Western Blotting\n",
      "3. Antibodies for Western Blotting\n",
      "4. Western Blotting: Products, Protocols, & Applications\n",
      "5. Tech Tip: Optimize Antigen and Antibody Concentrations for Western Blots (Page: 3, 4)\n",
      "6. FAQ:What is the recommended dilution range for SuperSignal West Pico PLUS Chemiluminescent Substrate?\n",
      "\n",
      "7. 5 Steps to Multiplexed Fluorescent Western Blotting\n",
      "NP00061 NuPAGE® Novex® Pre-Cast Gel Formulations\n",
      "The NuPAGE® Novex® Pre-Cast Gels are available in different acrylamide concentrations. The formulations are as follows:\n",
      "Gel Type Formulation Stacking Gel Separating Gel pH\n",
      "NuPAGE® Novex® Bis-Tris Gels Bis-Tris-HCl buffer (pH 6.4), Acrylamide, Bis-acrylamide, APS, Ultrapure water 4% 10%, 12%, 4–12% 7.0\n",
      "NuPAGE® Novex® Tris-Acetate Gels Tris-base, Acetic acid, Acrylamide, Bis-acrylamide, TEMED, APS, Ultrapure water 3.2% 7%, 3–8% 8.1\n",
      "Sources 1\n",
      "Additional Data & References\n",
      "1. NuPAGE Technical Guide (Page: 3, 5, 8, 9, 11, 12, 27)\n",
      "2. NuPAGE Large Protein Analysis System (Page: 9) 34577\n",
      "modified_url\n",
      "https://www.thermofisher.com/order/catalog/product/NP00061?SID=srch-srp-NP00061 NP00061\n",
      "NuPAGE\u00174 Novex\u00174 Pre-Cast Gel Formulations\n",
      "The NuPAGE\u00174 Novex\u00174 Pre-Cast Gels are available in two main types with different acrylamide concentrations and buffer systems:\n",
      "Gel Type Formulation Stacking Gel Separating Gel pH\n",
      "NuPAGE\u00174 Novex\u00174 Bis-Tris Gels Bis-Tris-HCl buffer (pH 6.4), Acrylamide, Bis-acrylamide, APS, Ultrapure water 4% 10%, 12%, 4\u001312% 7.0\n",
      "NuPAGE\u00174 Novex\u00174 Tris-Acetate Gels Tris-base, Acetic acid, Acrylamide, Bis-acrylamide, TEMED, APS, Ultrapure water 3.2% 7%, 3\u00138% 8.1\n",
      "Additional Details: The crosslinker concentration ranges from 3.8% to 5% depending on the gel region.\n",
      "Sources 1\n",
      "Additional Data & References\n",
      "1. NuPAGE Technical Guide (Page: 3, 5, 8, 9, 11, 12, 27)\n",
      "2. NuPAGE Large Protein Analysis System (Page: 9)\n",
      "NP0006 Hazard Information for NuPAGE Transfer Buffer (20X)\n",
      "Warning: This product contains a chemical (acrylamide) known to the state of California to cause cancer. Gloves should be worn at all times when handling gels.\n",
      "Refer to the Safety Data Sheet (SDS) for detailed safety information.\n",
      "\n",
      "\n",
      "Safety Data Sheet (1 matches)\n",
      "NP0006\n",
      "Sources 1\n",
      "Additional Data & References\n",
      "1. NuPAGE Technical Guide (Page: 1, 5, 48, 56)\n",
      "2. NuPAGE Large Protein Analysis System (Page: 4, 32)\n",
      "3. Western Blot Sample Preparation Protocol\n",
      "4. Western Blot Protocols and Recipes NP00061\n",
      "modified_url\n",
      "https://www.thermofisher.com/order/catalog/product/NP0006?SID=srch-srp-NP0006 NP0006\n",
      "Hazard Information for NuPAGE Transfer Buffer (20X) (Catalog Number: np0006)\n",
      "The provided product information does not explicitly state that NuPAGE Transfer Buffer (20X) is hazardous. It is intended for research use only and not for diagnostic procedures.\n",
      "However, it is important to note that some related NuPAGE products, such as the NuPAGE Precast Gels, contain residual acrylamide monomer, which is a chemical known to the state of California to cause cancer. While this warning is specific to gels, it is advisable to handle all reagents with appropriate safety precautions.\n",
      "For detailed safety information, including hazards, handling, and protective measures, please refer to the Safety Data Sheet (SDS) for NuPAGE Transfer Buffer (20X).\n",
      "Refer to the Safety Data Sheet (SDS) for detailed safety information.\n",
      "\n",
      "\n",
      "Safety Data Sheet (1 matches)\n",
      "NP0006\n",
      "Sources 1\n",
      "Additional Data & References\n",
      "1. NuPAGE Technical Guide (Page: 1, 5, 48)\n",
      "2. NuPAGE Large Protein Analysis System (Page: 4, 13, 32)\n",
      "3. Western Blot Protocols and Recipes\n",
      "4. Western Blot Sample Preparation Protocol\n",
      "B153 I was unable to find specific information on what to do if bis-ANS gets in your eyes. Refer to the Safety Data Sheet (SDS) for detailed safety information. Please contact our Technical Support Team.\n",
      "\n",
      "Safety Data Sheet (1 matches)\n",
      "B153\n",
      "Sources i\n",
      "Additional Data & References\n",
      "1. Conformational properties of substrate proteins bound to a molecular chaperone alpha-crystallin.\n",
      "2. Other Nonpolar and Amphiphilic Probes—Section 13.5\n",
      "3. Probes for Tubulin and Other Cytoskeletal Proteins—Section 11.2\n",
      "4. Binding of recrystallized and chromatographically purified 8-anilino-1-naphthalenesulfonate to Escherichia coli lac repressor.\n",
      "5. Temperature-induced exposure of hydrophobic surfaces and its effect on the chaperone activity of alpha-crystallin.\n",
      "6. Recoverin is a zinc-binding protein.\n",
      "7. Interaction of 1,1'-bi(4-anilino)naphthalene-5,5'-disulfonic acid with alpha-crystallin.\n",
      "8. Heat-induced conformational change of human lens recombinant alphaA- and alphaB-crystallins.\n",
      "How can we help?\n",
      "Customer Service\n",
      "800 955 6288\n",
      "Customercare@thermofisher.com\n",
      "Technical Support\n",
      "800 955 6288\n",
      "Contact Technical Support\n",
      "Instrument Service\n",
      "800 955 6288\n",
      "Contact Instrument Service\n",
      "Sales\n",
      "Live Chat NP0006\n",
      "modified_url\n",
      "https://www.thermofisher.com/order/catalog/product/B153?SID=srch-srp-B153 B153\n",
      "Safety Information for Bis-ANS (Catalog Number: B153)\n",
      "The provided product information does not include specific first aid instructions for accidental eye exposure to Bis-ANS. For safety-related concerns such as accidental contact with eyes, it is important to follow general laboratory safety protocols: immediately rinse the eyes thoroughly with plenty of water for several minutes and seek medical attention if irritation persists.\n",
      "Refer to the Safety Data Sheet (SDS) for detailed safety information regarding handling, first aid measures, and protective measures for Bis-ANS.\n",
      "If you require further assistance, please contact our Technical Support Team.\n",
      "\n",
      "\n",
      "Safety Data Sheet (1 matches)\n",
      "B153\n",
      "Additional Data & References\n",
      "1. Probes for Tubulin and Other Cytoskeletal Proteins—Section 11.2\n",
      "2. Other Nonpolar and Amphiphilic Probes—Section 13.5\n",
      "3. Binding of recrystallized and chromatographically purified 8-anilino-1-naphthalenesulfonate to Escherichia coli lac repressor.\n",
      "4. The kinetics of the acid-induced conformational change of influenza virus haemagglutinin can be followed using 1,1'-bis(4-anilino-5-naphthalenesulphonic acid).\n",
      "5. Mapping fatty acid binding to beta-lactoglobulin: Ligand binding is restricted by modification of Cys 121.\n",
      "6. Conformational properties of substrate proteins bound to a molecular chaperone alpha-crystallin.\n",
      "7. Role of the C-terminal extensions of alpha-crystallins. Swapping the C-terminal extension of alpha-crystallin to alphaB-crystallin results in enhanced chaperone activity.\n",
      "8. Temperature-induced exposure of hydrophobic surfaces and its effect on the chaperone activity of alpha-crystallin.\n",
      "How can we help?\n",
      "Customer Service\n",
      "800 955 6288\n",
      "Customercare@thermofisher.com\n",
      "Technical Support\n",
      "800 955 6288\n",
      "Contact Technical Support\n",
      "Instrument Service\n",
      "800 955 6288\n",
      "Contact Instrument Service\n",
      "Sales\n",
      "Live Chat\n",
      "B153 Consumption of Bis-ANS\n",
      "Bis-ANS is intended for research use only and is not for use in diagnostic procedures. It should not be consumed or ingested under any circumstances.\n",
      "Refer to the Safety Data Sheet (SDS) for detailed safety information.\n",
      "\n",
      "\n",
      "Safety Data Sheet (1 matches)\n",
      "B153\n",
      "Additional Data & References\n",
      "1. Probes for Tubulin and Other Cytoskeletal Proteins—Section 11.2\n",
      "2. Other Nonpolar and Amphiphilic Probes—Section 13.5\n",
      "3. Proteolytic digestion of alpha-lactalbumin: physiological implications.\n",
      "4. Spare Parts List for the SOLA II And SOLA II Trace Sulfur Online Analyzer (Page: 4)\n",
      "5. cDNA cloning and expression of the human hepatic bile acid-binding protein. A member of the monomeric reductase gene family.\n",
      "6. Selective adsorption of bis(1-anilino-8-naphthalenesulfonate) to the multiple forms of lactic dehydrogenase.\n",
      "7. 4,4'-Bis[8-(phenylamino)naphthalene-1-sulfonate] binding to human thrombins: a sensitive exo site fluorescent affinity probe.\n",
      "8. Aluminum fluoride interactions with troponin C. B153\n",
      "inside else\n",
      "Consumption of Bis-ANS (Catalog Number: B153)\n",
      "Bis-ANS is intended for research use only and is not for use in diagnostic procedures or for consumption. It is a chemical probe used primarily as a protein folding indicator and should be handled with appropriate laboratory safety precautions.\n",
      "Consumption of this product is not safe and should be strictly avoided.\n",
      "Refer to the Safety Data Sheet (SDS) for detailed safety information regarding handling, potential hazards, and first aid measures.\n",
      "If you have any concerns or require further assistance, please contact our Technical Support Team.\n",
      "\n",
      "\n",
      "Safety Data Sheet (1 matches)\n",
      "B153\n",
      "Additional Data & References\n",
      "1. Probes for Tubulin and Other Cytoskeletal Proteins—Section 11.2\n",
      "2. Other Nonpolar and Amphiphilic Probes—Section 13.5\n",
      "3. Stability studies on a lipase from Bacillus subtilis in guanidinium chloride.\n",
      "4. BisANS binding to tubulin: isothermal titration calorimetry and the site-specific proteolysis reveal the GTP-induced structural stability of tubulin.\n",
      "5. Proteolytic digestion of alpha-lactalbumin: physiological implications.\n",
      "6. The kinetics of the acid-induced conformational change of influenza virus haemagglutinin can be followed using 1,1'-bis(4-anilino-5-naphthalenesulphonic acid).\n",
      "7. Urea-induced denaturation of beta-trypsin: an evidence for a molten globule state.\n",
      "8. Binding of recrystallized and chromatographically purified 8-anilino-1-naphthalenesulfonate to Escherichia coli lac repressor.\n",
      "How can we help?\n",
      "Customer Service\n",
      "800 955 6288\n",
      "Customercare@thermofisher.com\n",
      "Technical Support\n",
      "800 955 6288\n",
      "Contact Technical Support\n",
      "Instrument Service\n",
      "800 955 6288\n",
      "Contact Instrument Service\n",
      "Sales\n",
      "Live Chat\n",
      "NP0007 Optimal Run Times for Gels Using MES and MOPS Running Buffers\n",
      "For NuPAGE Bis-Tris Gels:\n",
      "MES Running Buffer: Run for 35 minutes at 200 V constant.\n",
      "MOPS Running Buffer: Run for 50 minutes at 200 V constant.\n",
      "Sources 3\n",
      "4\n",
      "Additional Data & References\n",
      "1. NuPAGE Bis-Tris and Bolt Bis-Tris Plus Gels\n",
      "2. Quick Reference: NuPAGE Bis-Tris Mini Gels (Page: 2, 2)\n",
      "4. NuPAGE Technical Guide (Page: 9, 13, 18, 46)\n",
      "5. Protein Electrophoresis Buffers and Reagents B153\n",
      "modified_url\n",
      "https://www.thermofisher.com/order/catalog/product/NP0007?SID=srch-srp-NP0007 NP0007\n",
      "Optimal Run Times for Gels Using MES and MOPS Running Buffers\n",
      "When performing denaturing protein gel electrophoresis with NuPAGE Bis-Tris Mini Gels:\n",
      "Using MES Running Buffer: Run the gel for 35 minutes at 200 V constant.\n",
      "Using MOPS Running Buffer: Run the gel for 50 minutes at 200 V constant.\n",
      "These run times provide optimal protein separation depending on the gel percentage and power supply used.\n",
      "Additionally, an accelerated protocol called the NuPAGE Turbo Protocol allows excellent separation in only 25 minutes when using NuPAGE Novex Bis-Tris Gels with MES SDS Running Buffer at 250 V constant.\n",
      "References\n",
      "These recommendations are based on the NuPAGE Bis-Tris Mini Gel protocols and electrophoresis conditions.\n",
      "Sources 3\n",
      "4\n",
      "Additional Data & References\n",
      "1. NuPAGE Bis-Tris and Bolt Bis-Tris Plus Gels\n",
      "2. Quick Reference: NuPAGE Bis-Tris Mini Gels (Page: 2, 2)\n",
      "4. NuPAGE Technical Guide (Page: 13, 18, 46)\n",
      "5. Protein Electrophoresis Buffers and Reagents\n",
      "6. Protein Electrophoresis Buffers and Reagents\n",
      "l7526 Fixing Cells After Staining with LysoTracker Green DND-26\n",
      "LysoTracker Green DND-26 is a non-fixable dye, meaning it is not suitable for fixing cells after staining. This dye is designed for live-cell staining and tracking of acidic organelles such as lysosomes.\n",
      "For more information, you can visit our Cell Analysis Support Center.\n",
      "Additional Data & References\n",
      "1. Probes for Lysosomes, Peroxisomes and Yeast Vacuoles—Section 12.3\n",
      "2. LysoTracker and LysoSensor Probes (Page: 1, 2, 3, 4)\n",
      "3. FAQ:What is the recommended incubation time for LysoTracker Green DND-26 (Cat. No. L7526) and LysoSensor Yellow/Blue DND-160 (Cat. No. L7545)? NP0007\n",
      "modified_url\n",
      "https://www.thermofisher.com/order/catalog/product/l7526?SID=srch-srp-l7526 l7526\n",
      "Fixation of Cells After Staining with LysoTracker Green DND-26\n",
      "LysoTracker Green DND-26 is a cell-permeable, non-fixable dye that stains acidic compartments such as lysosomes. According to the product information, this dye is non-fixable, meaning that the staining is not retained after fixation of the cells. Therefore, fixing cells after staining with LysoTracker Green DND-26 is generally not recommended if you want to preserve the fluorescent labeling of acidic organelles.\n",
      "However, it is noted that some LysoTracker probes, such as LysoTracker Red DND-99, usually retain their staining pattern following fixation with aldehydes, but this property does not apply to LysoTracker Green DND-26.\n",
      "In summary, for LysoTracker Green DND-26, the staining is best visualized in live cells prior to fixation, as fixation will likely result in loss of fluorescence signal.\n",
      "Sources 1\n",
      "Additional Data & References\n",
      "1. Probes for Lysosomes, Peroxisomes and Yeast Vacuoles—Section 12.3\n",
      "2. LysoTracker and LysoSensor Probes (Page: 1, 3, 4)\n",
      "3. Live intracellular super-resolution imaging using site-specific stains.\n",
      "4. Effect of cryopreservation of bovine sperm organelle function and viability as determined by flow cytometry.\n",
      "5. Comparative pathogenesis of haloacetic acid and protein kinase inhibitor embryotoxicity in mouse whole embryo culture.\n",
      "6. Fluorometric assessments of acrosomal integrity and viability in cryopreserved bovine spermatozoa.\n",
      "NP0006 Using SimplyBlue™ SafeStain\n",
      "Introduction\n",
      "The NuPAGE® Large Protein Staining Kit includes the reagents to perform electrophoresis and staining with SimplyBlue™ SafeStain. SimplyBlue™ SafeStain is a ready-to-use, proprietary Coomassie G-250 stain that is specially formulated for fast, sensitive detection and safe, non-hazardous disposal. Proteins stained using SimplyBlue™ SafeStain are compatible with mass spectrometry (MS) analysis.\n",
      "Basic Protocol\n",
      "Rinse the mini-gel 3 times for 5 minutes with 100 mL deionized water to remove SDS and buffer salts, which interfere with binding of the dye to the protein. Discard each rinse.\n",
      "Stain the mini-gel with enough SimplyBlue™ SafeStain (~20 mL) to cover the mini-gel. Stain for 1 hour at room temperature with gentle shaking. Bands will begin to develop within minutes. After incubation, discard the stain. Do not re-use the stain.\n",
      "Wash the mini-gel with 100 mL water for 1-3 hours. The gel can be left in the water for several days without loss of sensitivity.\n",
      "Microwave Protocol\n",
      "After electrophoresis, place the gel in 100 mL deionized water in a loosely covered container and microwave on High (950 to 1100 watts) for 1 minute until the solution almost boils.\n",
      "Shake the gel on an orbital shaker for 1 minute. Discard the water.\n",
      "Repeat Steps 1-2 two more times.\n",
      "After the last wash, add 20 mL SimplyBlue™ SafeStain and microwave on High for 45 seconds to 1 minute until the solution almost boils.\n",
      "Shake the gel on an orbital shaker for 5 minutes. (Detection limit 20 ng BSA.)\n",
      "Wash the gel in 100 mL of ultrapure water for 10 minutes on a shaker. (Detection limit 10 ng BSA.)\n",
      "Add 20 mL of 20% NaCl for at least 5 minutes. (Detection limit 5 ng BSA.) The gel can be stored for several weeks in the salt solution.\n",
      "Disposal Information\n",
      "In most cases, the SimplyBlue™ SafeStain can be diluted 1:20 with water and flushed down the drain. Diluting the stain raises the pH to between 5 and 7, which complies with most federal and state regulations. For disposal requirements in your area, consult your safety officer.\n",
      "Drying the Gel\n",
      "The stained gel can be dried for storage or analysis by vacuum-drying or air-drying. We recommend using the DryEase™ Mini-Gel Drying System to air-dry the gel. When using the DryEase™ Mini-Gel Drying System, be sure to restrict the incubation time of your SimplyBlue™ SafeStain stained gel in the Gel-Dry™ Drying Solution to 5 minutes because longer exposure can result in destaining of bands.\n",
      "Sources 1\n",
      "Additional Data & References\n",
      "1. NuPAGE Large Protein Analysis System (Page: 8, 21, 22, 23, 24, 29)\n",
      "2. NuPAGE Technical Guide (Page: 24, 25) l7526\n",
      "modified_url\n",
      "https://www.thermofisher.com/order/catalog/product/NP0006?SID=srch-srp-NP0006 NP0006\n",
      "How to Use SimplyBlue\u000217 SafeStain\n",
      "Introduction\n",
      "SimplyBlue\u000217 SafeStain is a ready-to-use Coomassie G-250 stain included in the NuPAGE Large Protein Staining Kit. It is formulated for fast, sensitive protein detection and safe disposal. Proteins stained with SimplyBlue\u000217 SafeStain are compatible with mass spectrometry analysis.\n",
      "Basic Protocol\n",
      "Rinse the mini-gel 3 times for 5 minutes each with 100 mL deionized water to remove SDS and buffer salts. Discard each rinse.\n",
      "Stain the mini-gel with approximately 20 mL SimplyBlue\u000217 SafeStain to cover the gel. Incubate for 1 hour at room temperature with gentle shaking. Bands will develop within minutes. After staining, discard the stain; do not reuse.\n",
      "Wash the mini-gel with 100 mL water for 1-3 hours. The gel can remain in water for several days without loss of sensitivity.\n",
      "Notes:\n",
      "An alcohol/acetic acid fixing step is not required or recommended before staining.\n",
      "The gel can be stained up to 3 hours; beyond that, sensitivity decreases. For overnight staining, add 2 mL 20% NaCl per 20 mL stain to maintain sensitivity.\n",
      "For highest sensitivity, after 1 hour wash, add 20 mL 20% NaCl to fresh water and continue washing for 2 hours to overnight (detection limit: 7 ng BSA).\n",
      "For clearest background for photography, perform a second 1-hour wash with 100 mL water.\n",
      "After the second wash, sensitivity decreases if the gel stays in water more than 1 day; add 20 mL 20% NaCl if storing for a few days.\n",
      "Microwave Protocol (Faster Method)\n",
      "Place gel in 100 mL deionized water in a loosely covered container and microwave on High (950-1100 watts) for 1 minute until near boiling.\n",
      "Shake gel on orbital shaker for 1 minute; discard water.\n",
      "Repeat steps 1 and 2 two more times.\n",
      "Add 20 mL SimplyBlue\u000217 SafeStain and microwave on High for 45 seconds to 1 minute until near boiling.\n",
      "Shake gel on orbital shaker for 5 minutes (detection limit 20 ng BSA).\n",
      "Wash gel in 100 mL ultrapure water for 10 minutes on shaker (detection limit 10 ng BSA).\n",
      "Add 20 mL 20% NaCl for at least 5 minutes (detection limit 5 ng BSA). The gel can be stored for several weeks in this salt solution.\n",
      "Drying the Gel\n",
      "The stained gel can be dried for storage or analysis by vacuum-drying or air-drying. We recommend using the DryEase\u000217 Mini-Gel Drying System. When using this system, limit incubation in Gel-Dry\u000217 Drying Solution to 5 minutes to avoid destaining.\n",
      "Disposal\n",
      "SimplyBlue\u000217 SafeStain can usually be diluted 1:20 with water and flushed down the drain, raising the pH to 5-7 to comply with most regulations. Consult your safety officer for local disposal requirements.\n",
      "Sources 1\n",
      "2\n",
      "Additional Data & References\n",
      "1. NuPAGE Large Protein Analysis System (Page: 8, 21, 22, 23, 24, 29)\n",
      "2. NuPAGE Technical Guide (Page: 24, 25)\n",
      "NP0336BOX Polyacrylamide Concentrations of NuPAGE Bis-Tris Protein Gels\n",
      "The NuPAGE Bis-Tris protein gels come in four polyacrylamide concentrations:\n",
      "8%\n",
      "10%\n",
      "12%\n",
      "4-12% gradient\n",
      "Sources 1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "Additional Data & References\n",
      "1. Quick Reference: NuPAGE Bis-Tris Mini Gels (Page: 1)\n",
      "2. NuPAGE™ Bis-Tris Mini Protein Gels, 4–12%, 1.0–1.5 mm\n",
      "3. NuPAGE™ Bis-Tris Mini Protein Gels, 4–12%, 1.0–1.5 mm\n",
      "4. NuPAGE™ Bis-Tris Mini Protein Gels, 4–12%, 1.0–1.5 mm\n",
      "5. NuPAGE™ Bis-Tris Mini Protein Gels, 4–12%, 1.0–1.5 mm\n",
      "6. NuPAGE™ Bis-Tris Mini Protein Gels, 4–12%, 1.0–1.5 mm\n",
      "7. NuPAGE™ Bis-Tris Mini Protein Gels, 4–12%, 1.0–1.5 mm\n",
      "8. NuPAGE™ Bis-Tris Mini Protein Gels, 4–12%, 1.0–1.5 mm NP0006\n",
      "modified_url\n",
      "https://www.thermofisher.com/order/catalog/product/NP0336BOX?SID=srch-srp-NP0336BOX NP0336BOX\n",
      "Polyacrylamide Concentrations of NuPAGE Bis-Tris Protein Gels\n",
      "NuPAGE Bis-Tris protein gels are available in four polyacrylamide concentrations to provide optimal separation of proteins. These concentrations are:\n",
      "8%\n",
      "10%\n",
      "12%\n",
      "4 6% gradient\n",
      "These gels are formulated for denaturing gel electrophoresis applications and are designed to separate small- to medium-sized proteins ranging from 1.5 to 300 kDa.\n",
      "NuPAGE Bis-Tris gels come in mini (8 cm x 8 cm) or midi (8.7 cm x 13.3 cm) sizes and thicknesses of 1.0 mm (mini and midi) or 1.5 mm (mini only). Multiple well formats are also available.\n",
      "For optimal sample preparation, use NuPAGE LDS Sample Buffer and NuPAGE Sample Reducing Agent. Running buffers such as NuPAGE MES SDS Running Buffer are recommended for better resolution of small proteins, while NuPAGE MOPS SDS Running Buffer is suitable for medium- to large-size proteins.\n",
      "These details ensure you can select the right gel and running buffer combination to achieve the best protein separation results.\n",
      "Sources 4\n",
      "Additional Data & References\n",
      "1. Quick Reference: NuPAGE Bis-Tris Mini Gels (Page: 1, 1, 2)\n",
      "2. NuPAGE Bis-Tris and Bolt Bis-Tris Plus Gels\n",
      "3. One-Dimensional SDS Gel Electrophoresis of Proteins with NuPAGE® Novex® Pre-Cast Gels\n",
      "5. NuPAGE Technical Guide (Page: 6, 8)\n",
      "6. FAQ:Does the NuPAGE 4 to 12%, Bis-Tris, 1.0–1.5 mm, Mini Protein Gels (Cat. No. NP0324BOX) have an extra well to load DNA ladder?\n",
      "NP0002 Solubility Information for NuPAGE MES SDS Running Buffer (20X)\n",
      "The NuPAGE MES SDS Running Buffer (20X) is soluble in water.\n",
      "Additional Data & References\n",
      "1. Quick Reference: NuPAGE Bis-Tris Mini Gels (Page: 1, 1, 3, 3)\n",
      "3. NuPAGE Bis-Tris and Bolt Bis-Tris Plus Gels\n",
      "4. Protein Electrophoresis Buffers and Reagents\n",
      "5. FAQ:Can I use NuPAGE Bis-Tris gels with NuPAGE MOPS or MES Running Buffer prepared without SDS for electrophoresis under native conditions?\n",
      "\n",
      "6. FAQ:Are the NuPAGE MOPS and MES SDS running buffers compatible with N-terminal sequencing of proteins via Edman degradation? NP0336BOX\n",
      "modified_url\n",
      "https://www.thermofisher.com/order/catalog/product/NP0002?SID=srch-srp-NP0002 NP0002\n",
      "Solubility Information for NuPAGE MES SDS Running Buffer (20X) - np0002\n",
      "The NuPAGE MES SDS Running Buffer (20X) is soluble in water.\n",
      "Additional Data & References\n",
      "1. Quick Reference: NuPAGE Bis-Tris Mini Gels (Page: 1, 1, 3, 3)\n",
      "3. NuPAGE Bis-Tris and Bolt Bis-Tris Plus Gels\n",
      "4. Protein Electrophoresis Buffers and Reagents\n",
      "5. Protein Electrophoresis Buffers and Reagents\n",
      "6. FAQ:Can I use NuPAGE Bis-Tris gels with NuPAGE MOPS or MES Running Buffer prepared without SDS for electrophoresis under native conditions?\n",
      "a53815 Razón para amplificar el ADN donante\n",
      "La amplificación del ADN donante es necesaria para generar una cantidad suficiente de ADN de doble cadena lineal (dsDNA) con los brazos de homología específicos del gen deseado. Este proceso asegura que haya suficiente material genético para la transfección y la edición del genoma.\n",
      "El protocolo de amplificación del ADN donante TrueTag™ incluye la configuración de reacciones de PCR y la confirmación de la reacción exitosa mediante electroforesis en gel de agarosa. Esto garantiza que el ADN donante tenga el tamaño de banda correcto para diferentes tipos de donantes TrueTag™.\n",
      "Para más detalles, consulte la documentación del producto.\n",
      "Sources 1\n",
      "2\n",
      "Datos Adicionales y Referencias\n",
      "1. User Guide: TrueTag DNA Donor Kit (Page: 3, 27, 37, 38, 40, 51)\n",
      "2. Flyer: True story: There's an easier way to tag and enrich cells (Page: 2)\n",
      "3. Workflow guide for using the TrueTag Knockout Enrichment Kit for knockout cell line generation (Page: 4) NP0002\n",
      "modified_url\n",
      "https://www.thermofisher.com/order/catalog/product/a53815?SID=srch-srp-a53815 a53815\n",
      "¿Por qué es necesario amplificar el donante de ADN?\n",
      "La amplificación por PCR del donante de ADN es un paso esencial para generar una cantidad suficiente de ADN lineal de doble cadena con los brazos de homología específicos del gen deseado. Esto permite obtener más de 5 b5g de ADN donante listo para la transfeccif3n y la edicif3n genf3mica.\n",
      "La amplificacif3n tambie9n permite agregar brazos de homologeda a la plantilla de donante mediante cebadores especedficos, lo que es crucial para la reparacif3n dirigida por homologa (HDR) y la insercif3n precisa del donante en el genoma objetivo.\n",
      "Ademe1s, la amplificacif3n elimina la necesidad de pasos de clonacif3n laboriosos y prolongados, reduciendo el tiempo total del protocolo a unas pocas horas y aumentando la eficiencia de la edicif3n.\n",
      "En resumen, amplificar el donante de ADN mediante PCR:\n",
      "Proporciona suficiente cantidad de ADN para la transfeccif3n.\n",
      "Permite la adicif3n de brazos de homologa para la insercif3n dirigida.\n",
      "Reduce el tiempo y la complejidad del protocolo al eliminar pasos de clonacif3n.\n",
      "Este proceso este1 detallado en el protocolo del kit TrueTag Donor DNA, que incluye la mezcla maestra de PCR y los cebadores necesarios para la amplificacif3n.\n",
      "Sources 1\n",
      "Datos Adicionales y Referencias\n",
      "1. User Guide: TrueTag DNA Donor Kit (Page: 15, 27, 37, 38, 40)\n",
      "2. Flyer: True story: There's an easier way to tag and enrich cells (Page: 2)\n",
      "3. Workflow guide for using the TrueTag Knockout Enrichment Kit for knockout cell line generation (Page: 4)\n",
      "4. Knock-in Editing and Gene Tagging\n",
      "QGP-306-RATHKPANEL Funktionen der QuantiGene Plex Panels\n",
      "Die QuantiGene Plex Panels bieten mehrere Funktionen, die für die Genexpressionsanalyse nützlich sind:\n",
      "Echte Multiplexing-Fähigkeit — Messen Sie mehrere Gene von Interesse und Haushaltsgene im selben Well ohne Kreuzreaktivität.\n",
      "Anpassung an schwierige Probentypen — Funktioniert mit degradierten und vernetzten RNA in FFPE-Geweben und Blut.\n",
      "Standardisierte Plattform — Verwendet ein 96-Well-Format, das mit Luminex 200, FLEXMAP 3D und INTELLIFLEX-Systemen kompatibel ist (384-Well-Format auf Anfrage verfügbar).\n",
      "Einfacher Workflow — Verwendet einen ELISA-ähnlichen Workflow mit direkter Hybridisierung von Transkripten an Beads und Transkriptmarkierung (branched DNA-Signalamplifikation und PE-Auslesung).\n",
      "Anpassung mit validierten Zielen — Wenn alternative Ziele in diesem Panel benötigt werden, wählen Sie aus unserem großen Inventar von über 22.000 Genen, um benutzerdefinierte Forschungspanels zu erstellen.\n",
      "Schnelle Durchlaufzeit — Dieses Katalogpanel ermöglicht eine schnelle Lieferung.\n",
      "Ziele: Actb, Gapdh, Gusb, Hprt, Tbp, Ppib\n",
      "Panel- und Signalamplifikationstechnologie\n",
      "QuantiGene Plex Assays sind Hybridisierungs-basierte Assays zur direkten Quantifizierung von RNA direkt aus Zelllysaten, Gewebehomogenaten oder gereinigter RNA. Sie ermöglichen die Messung mehrerer Ziele im selben Well durch Multiplexing. QuantiGene Plex Panels enthalten zwei Komponenten: ein zielgerichtetes Sondenset und magnetische Fangbeads. Die zielgerichteten Sonden werden verwendet, um die interessierende RNA zu erfassen, und die branched DNA-Technologie verstärkt dann die Fluoreszenzsignale, die mit einem Luminex xMAP-Instrument ausgelesen werden. Die Intensität des Fluoreszenzsignals ist direkt proportional zur Menge an Ziel-mRNA, die in der Probe vorhanden ist. Dies ermöglicht die (relative) Quantifizierung der Genexpression und mit diesem spezifischen Panel die menschliche Immunprofilierung. Eine detaillierte und schrittweise Datenanalyse wird durch die QuantiGene Plex Datenanalyse-App erleichtert.\n",
      "Anwendungen\n",
      "Forschungsanwendungen des QuantiGene Plex Assays umfassen das Screening von Wirkstoffkandidaten, die Verifizierung von Biomarkern, die Bestimmung der siRNA-Knockdown-Effizienz, die prospektive und/oder retrospektive Analyse klinischer Proben, die Validierung von Mikroarrays, die prädiktive Toxikologieforschung und die Erkennung von Translokations-/Fusionsgenen. Im Allgemeinen kann jede Multiplex-Genexpressionsanalyse mit QuantiGene Plex Assays durchgeführt werden, unabhängig von Spezies oder Probentyp.\n",
      "Zusätzliche Daten & Referenzen\n",
      "1. User Guide: QuantiGene Plex Gene Expression Assay (Page: 3, 5, 9)\n",
      "2. User Guide: QuantiGene Plex Gene Expression Assay (Page: 1, 3, 5, 8, 32) a53815\n",
      "modified_url\n",
      "https://www.thermofisher.com/order/catalog/product/QGP-306-RATHKPANEL?SID=srch-srp-QGP-306-RATHKPANEL QGP-306-RATHKPANEL\n"
     ]
    },
    {
     "ename": "NoSuchElementException",
     "evalue": "Message: no such element: Unable to locate element: {\"method\":\"xpath\",\"selector\":\"/html/body/div[10]/div/div[3]/div[2]/core-navigationbar/core-navigationbar-item[2]//div\"}\n  (Session info: chrome=143.0.7499.147); For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#nosuchelementexception\nStacktrace:\n0   chromedriver                        0x0000000104ad8b2c cxxbridge1$str$ptr + 3033596\n1   chromedriver                        0x0000000104ad0b30 cxxbridge1$str$ptr + 3000832\n2   chromedriver                        0x00000001045cab0c _RNvCsgXDX2mvAJAg_7___rustc35___rust_no_alloc_shim_is_unstable_v2 + 74196\n3   chromedriver                        0x00000001046119c8 _RNvCsgXDX2mvAJAg_7___rustc35___rust_no_alloc_shim_is_unstable_v2 + 364688\n4   chromedriver                        0x0000000104652968 _RNvCsgXDX2mvAJAg_7___rustc35___rust_no_alloc_shim_is_unstable_v2 + 630832\n5   chromedriver                        0x000000010460616c _RNvCsgXDX2mvAJAg_7___rustc35___rust_no_alloc_shim_is_unstable_v2 + 317492\n6   chromedriver                        0x0000000104a9c098 cxxbridge1$str$ptr + 2785128\n7   chromedriver                        0x0000000104a9f804 cxxbridge1$str$ptr + 2799316\n8   chromedriver                        0x0000000104a7c850 cxxbridge1$str$ptr + 2656032\n9   chromedriver                        0x0000000104aa0074 cxxbridge1$str$ptr + 2801476\n10  chromedriver                        0x0000000104a6d1cc cxxbridge1$str$ptr + 2592924\n11  chromedriver                        0x0000000104abfcb4 cxxbridge1$str$ptr + 2931588\n12  chromedriver                        0x0000000104abfe34 cxxbridge1$str$ptr + 2931972\n13  chromedriver                        0x0000000104ad0788 cxxbridge1$str$ptr + 2999896\n14  libsystem_pthread.dylib             0x0000000183f2bbc8 _pthread_start + 136\n15  libsystem_pthread.dylib             0x0000000183f26b80 thread_start + 8\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNoSuchElementException\u001b[0m                    Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[55], line 16\u001b[0m\n\u001b[1;32m     14\u001b[0m  time\u001b[38;5;241m.\u001b[39msleep(\u001b[38;5;241m15\u001b[39m)\n\u001b[1;32m     15\u001b[0m  \u001b[38;5;66;03m# Click on the specific tab before entering the input text #/html/body/div[10]/div/div[3]/div[2]/core-navigationbar/core-navigationbar-item[2]//div\u001b[39;00m\n\u001b[0;32m---> 16\u001b[0m  segment_tab \u001b[38;5;241m=\u001b[39m driver\u001b[38;5;241m.\u001b[39mfind_element(By\u001b[38;5;241m.\u001b[39mXPATH, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/html/body/div[10]/div/div[3]/div[2]/core-navigationbar/core-navigationbar-item[2]//div\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;66;03m#\u001b[39;00m\n\u001b[1;32m     18\u001b[0m \u001b[38;5;66;03m# segment_tab.click()     #/html/body/div[10]/div/div[2]/div[2]/core-navigationbar/core-navigationbar-item[2]//div                   \u001b[39;00m\n\u001b[1;32m     19\u001b[0m  \u001b[38;5;66;03m#/html/body/div[9]/div/div[3]/div[2]/core-navigationbar/core-navigationbar-item[2]//div\u001b[39;00m\n\u001b[1;32m     20\u001b[0m  \u001b[38;5;66;03m# /html/body/div[9]/div/div[2]/div[2]/core-navigationbar/core-navigationbar-item[2]//div/a \u001b[39;00m\n\u001b[1;32m     21\u001b[0m  \u001b[38;5;66;03m# Wait for the tab to load (adjust the sleep time as needed) #/html/body/div[9]/div/div[3]/div[2]/core-navigationbar/core-navigationbar-item[2]//div/a\u001b[39;00m\n\u001b[1;32m     22\u001b[0m  time\u001b[38;5;241m.\u001b[39msleep(\u001b[38;5;241m2\u001b[39m)\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py:926\u001b[0m, in \u001b[0;36mWebDriver.find_element\u001b[0;34m(self, by, value)\u001b[0m\n\u001b[1;32m    923\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m NoSuchElementException(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot locate relative element with: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mby\u001b[38;5;241m.\u001b[39mroot\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    924\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m elements[\u001b[38;5;241m0\u001b[39m]\n\u001b[0;32m--> 926\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mexecute(Command\u001b[38;5;241m.\u001b[39mFIND_ELEMENT, {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124musing\u001b[39m\u001b[38;5;124m\"\u001b[39m: by, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvalue\u001b[39m\u001b[38;5;124m\"\u001b[39m: value})[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvalue\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py:458\u001b[0m, in \u001b[0;36mWebDriver.execute\u001b[0;34m(self, driver_command, params)\u001b[0m\n\u001b[1;32m    455\u001b[0m response \u001b[38;5;241m=\u001b[39m cast(RemoteConnection, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcommand_executor)\u001b[38;5;241m.\u001b[39mexecute(driver_command, params)\n\u001b[1;32m    457\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m response:\n\u001b[0;32m--> 458\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39merror_handler\u001b[38;5;241m.\u001b[39mcheck_response(response)\n\u001b[1;32m    459\u001b[0m     response[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvalue\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_unwrap_value(response\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvalue\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    460\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.12/site-packages/selenium/webdriver/remote/errorhandler.py:232\u001b[0m, in \u001b[0;36mErrorHandler.check_response\u001b[0;34m(self, response)\u001b[0m\n\u001b[1;32m    230\u001b[0m         alert_text \u001b[38;5;241m=\u001b[39m value[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124malert\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtext\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    231\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exception_class(message, screen, stacktrace, alert_text)  \u001b[38;5;66;03m# type: ignore[call-arg]  # mypy is not smart enough here\u001b[39;00m\n\u001b[0;32m--> 232\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m exception_class(message, screen, stacktrace)\n",
      "\u001b[0;31mNoSuchElementException\u001b[0m: Message: no such element: Unable to locate element: {\"method\":\"xpath\",\"selector\":\"/html/body/div[10]/div/div[3]/div[2]/core-navigationbar/core-navigationbar-item[2]//div\"}\n  (Session info: chrome=143.0.7499.147); For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#nosuchelementexception\nStacktrace:\n0   chromedriver                        0x0000000104ad8b2c cxxbridge1$str$ptr + 3033596\n1   chromedriver                        0x0000000104ad0b30 cxxbridge1$str$ptr + 3000832\n2   chromedriver                        0x00000001045cab0c _RNvCsgXDX2mvAJAg_7___rustc35___rust_no_alloc_shim_is_unstable_v2 + 74196\n3   chromedriver                        0x00000001046119c8 _RNvCsgXDX2mvAJAg_7___rustc35___rust_no_alloc_shim_is_unstable_v2 + 364688\n4   chromedriver                        0x0000000104652968 _RNvCsgXDX2mvAJAg_7___rustc35___rust_no_alloc_shim_is_unstable_v2 + 630832\n5   chromedriver                        0x000000010460616c _RNvCsgXDX2mvAJAg_7___rustc35___rust_no_alloc_shim_is_unstable_v2 + 317492\n6   chromedriver                        0x0000000104a9c098 cxxbridge1$str$ptr + 2785128\n7   chromedriver                        0x0000000104a9f804 cxxbridge1$str$ptr + 2799316\n8   chromedriver                        0x0000000104a7c850 cxxbridge1$str$ptr + 2656032\n9   chromedriver                        0x0000000104aa0074 cxxbridge1$str$ptr + 2801476\n10  chromedriver                        0x0000000104a6d1cc cxxbridge1$str$ptr + 2592924\n11  chromedriver                        0x0000000104abfcb4 cxxbridge1$str$ptr + 2931588\n12  chromedriver                        0x0000000104abfe34 cxxbridge1$str$ptr + 2931972\n13  chromedriver                        0x0000000104ad0788 cxxbridge1$str$ptr + 2999896\n14  libsystem_pthread.dylib             0x0000000183f2bbc8 _pthread_start + 136\n15  libsystem_pthread.dylib             0x0000000183f26b80 thread_start + 8\n"
     ]
    }
   ],
   "source": [
    "#To be included later!!!\n",
    "# Iterate over the rows in the DataFrame\n",
    "for index, row in df_sheet2.iterrows():\n",
    "    SKU = row['SKU']\n",
    "    Question = row['Question']\n",
    "    expected_response = row['Expected Answer']\n",
    "    print (SKU, expected_response, previous_url_modifier)\n",
    "\n",
    "    # Modify the URL only if the URL modifier value changes\n",
    "    if SKU != previous_url_modifier:\n",
    "        modified_url = modify_url(base_url, SKU)\n",
    "        previous_url_modifier = SKU\n",
    "        driver.get(modified_url)\n",
    "        time.sleep(15)\n",
    "        # Click on the specific tab before entering the input text #/html/body/div[10]/div/div[3]/div[2]/core-navigationbar/core-navigationbar-item[2]//div\n",
    "        segment_tab = driver.find_element(By.XPATH, '/html/body/div[10]/div/div[3]/div[2]/core-navigationbar/core-navigationbar-item[2]//div') #\n",
    "                                                    \n",
    "       # segment_tab.click()     #/html/body/div[10]/div/div[2]/div[2]/core-navigationbar/core-navigationbar-item[2]//div                   \n",
    "        #/html/body/div[9]/div/div[3]/div[2]/core-navigationbar/core-navigationbar-item[2]//div\n",
    "        # /html/body/div[9]/div/div[2]/div[2]/core-navigationbar/core-navigationbar-item[2]//div/a \n",
    "        # Wait for the tab to load (adjust the sleep time as needed) #/html/body/div[9]/div/div[3]/div[2]/core-navigationbar/core-navigationbar-item[2]//div/a\n",
    "        time.sleep(2)\n",
    "    else:\n",
    "        driver.get(modified_url)\n",
    "        print (\"inside else\")\n",
    "        time.sleep(5)\n",
    "\n",
    "    # Find the input field and enter the text\n",
    "    input_field = driver.find_element(By.CSS_SELECTOR, '#search-component-wrapper-ssw > core-card > div > div.input-search-wrapper-ssw > core-search')  # Adjust the locator as needed\n",
    "    input_field.click()\n",
    "   # input_field.clear()\n",
    "    input_field.send_keys(Question)\n",
    "    input_field.send_keys(Keys.RETURN)\n",
    "\n",
    "    # Wait for the response to be generated (2 minutes)\n",
    "    time.sleep(50)\n",
    "\n",
    "    # Find the response field and get the text\n",
    "    #response_field = driver.find_element(By.CLASS_NAME, 'input-search hydrated')  # Adjust the locator as needed\n",
    "    #generated_response = response_field.text\n",
    "\n",
    "    response_field = driver.find_element(By.XPATH, '//*[@id=\"search-component-wrapper-ssw\"]/core-card/div/div[2]/div/div[2]/div/div[1]/div/div/div/div')  # Adjust the locator as needed\n",
    "    generated_response = response_field.text\n",
    "    print(generated_response)\n",
    "                                                    #//*[@id=\"search-component-wrapper-ssw\"]/core-card/div/div[2]/div/div[2]/div/div[1]/div/div/div/div\n",
    "                                                    #//*[@id=\"search-component-wrapper-ssw\"]/core-card/div/div[2]/div[2]/div[2]/div/div[1]/div/div/div/div\n",
    "                                                    #//*[@id=\"search-component-wrapper-ssw\"]/core-card/div/div[2]/div[2]/div[2]/div/div[1]/div/div/div\n",
    "                                                    #//*[@id=\"search-component-wrapper-ssw\"]/core-card/div/div[2]/div[3]/div[2]/div/div[1]/div\n",
    "    # Calculate cosine similarity\n",
    "    similarity = calculate_cosine_similarity(generated_response, expected_response)\n",
    "\n",
    "    # Compare the generated response with the expected response using cosine similarity\n",
    "    if similarity >= 0.60:  # Threshold for considering a match (adjust as needed)\n",
    "        matches.append((SKU,Question ,generated_response, expected_response, similarity))\n",
    "    else:\n",
    "        mismatches.append((SKU,Question ,generated_response, expected_response, similarity))\n",
    "\n",
    "# Close the web driver\n",
    "#driver.quit()\n",
    "\n",
    "print(matches)\n",
    "print(mismatches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47895b53-1666-4258-aee1-eaa21a017e0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "now=datetime.datetime.now()\n",
    "formatted_date_time = now.strftime(\"%Y-%m-%d %H:%M:%S\")\n",
    "print (\"current date time is \"+ str(formatted_date_time))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b662e559-76e2-4fa4-aec9-7519b9fd0bdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "matchingResults=\"matches_report_ProdPDP\"+str(now.strftime(\"%Y-%m-%d %H_%M_%S\"))+\".xlsx\"\n",
    "mismatchingResults=\"mismatches_report_ProdPDP\"+str(now.strftime(\"%Y-%m-%d %H_%M_%S\"))+\".xlsx\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "264d4522-b208-4264-94ea-b785878c9df3",
   "metadata": {},
   "outputs": [],
   "source": [
    "matchingResults, mismatchingResults"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66eeb046-f5de-41ba-a2ff-e37e30cf6c9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Generate the report\n",
    "matches_report = pd.DataFrame(matches, columns=[\"SKU\",\"Question\", 'Generated Response', 'Expected Response', 'Cosine Similarity'])\n",
    "#matches_report.to_excel(matchingResults, index=False)\n",
    "matches_report.to_csv(\"matches_report_ProdPDP.csv\", index=False)\n",
    "\n",
    "mismatches_report = pd.DataFrame(mismatches, columns=[\"SKU\",\"Question\", 'Generated Response', 'Expected Response', 'Cosine Similarity'])\n",
    "mismatches_report.to_excel(mismatchingResults, index=False)\n",
    "\n",
    "print(\"Report generated successfully.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88d8647f-2f81-4350-8440-43cdfb1e5ede",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
